cpp/test/mdarray.cu

```diff
@@ -436,6 +436,430 @@ TEST(MDArray, FuncArg)
+                         });
+      check_status(p_status, s);
+
+      // resulting submdspan is *NOT* padded anymore
+      static_assert(std::is_same_v<typename decltype(subspan)::layout_type, stdex::layout_stride>);
+    }
+
+    // sub-rectangle retaining padded layout
+    {
+      auto padded_span = padded_device_array.view();
+      auto subspan =
+        stdex::submdspan(padded_span, std::make_tuple(1ul, 4ul), std::make_tuple(2ul, 5ul));
+      thrust::for_each_n(rmm::exec_policy(s),
+                         thrust::make_counting_iterator(0ul),
+                         (rows - 1) * (cols - 2),
+                         [subspan, rows, cols, padded_span, p_status] __device__(size_t i) {
+                           size_t idx = i / (cols - 2);
+                           size_t idy = i % (cols - 2);
+                           // elements > subspan range can be accessed as well
+                           if (subspan(idx, idy) != padded_span(idx + 1, idy + 2))
+                             myAtomicAdd(p_status, 1);
+                         });
+      check_status(p_status, s);
+
+      // resulting submdspan is *NOT* padded anymore
+      static_assert(std::is_same_v<typename decltype(subspan)::layout_type, stdex::layout_stride>);
+    }
+  }
+}
+
+TEST(MDSpan, SubmdspanPadding) { test_submdspan_padding(); }*/
+
+struct TestElement1 {
+  int a, b;
+};
+
+void test_mdspan_padding_by_type()
+{
+  using extents_type = stdex::extents<size_t, dynamic_extent, dynamic_extent>;
+  auto s             = rmm::cuda_stream_default;
+
+  {
+    constexpr int rows            = 6;
+    constexpr int cols            = 7;
+    constexpr int alignment_bytes = 16;
+
+    thrust::device_vector<int32_t> status(1, 0);
+    auto p_status = status.data().get();
+
+    // manually check strides for row major (c style) padding
+    {
+      using padded_layout_row_major = detail::layout_padded_general<
+        detail::padding<std::remove_cv_t<std::remove_reference_t<TestElement1>>,
+                        alignment_bytes>::value,
+        StorageOrderType::row_major_t>;
+
+      using padded_mdarray_type =
+        device_mdarray<TestElement1, extents_type, padded_layout_row_major>;
+      auto device_policy = padded_mdarray_type::container_policy_type{s};
+
+      padded_layout_row_major::mapping<extents_type> layout{extents_type{rows, cols}};
+      padded_mdarray_type padded_device_array{layout, device_policy};
+      int alignment_elements = detail::padding<TestElement1, alignment_bytes>::value;
+      auto padded_span       = padded_device_array.view();
+      thrust::for_each_n(
+        rmm::exec_policy(s),
+        thrust::make_counting_iterator(0ul),
+        rows * cols,
+        [rows, cols, padded_span, alignment_elements, p_status] __device__(size_t i) {
+          size_t idx = i / cols;
+          size_t idy = i % cols;
+          if ((&(padded_span(idx, idy)) - &(padded_span(0, idy))) % alignment_elements != 0)
```

Comment:
So for a padded mdarray, does this require that we use the accessor directly and can't, for example, just use the underlying data_handle/pointer? Thinking about the case where we have a matrix and want it padded so we call down to something like cublas, which would be expecting a raw pointer.

Reply:
@cjnolet This might relate to how mdarray's design has changed.  The latest version of the mdarray proposal has a way to get an mdspan viewing the mdarray's elements, with a given accessor.

mdarray always has a container underneath, so you should always be able to use the mdarray's pointer directly.  On the other hand, a much nicer way to access the mdarray's elements on device would be to get an mdspan viewing the elements.